src/Features/Core/Portable/EditAndContinue/DeclarationBody.cs

```diff
@@ -16,24 +16,39 @@ internal abstract class DeclarationBody : IEquatable<DeclarationBody>
 
     /// <summary>
     /// Root nodes of the body. Descendant nodes of these roots include all nodes of the body and no nodes that do not belong to the body.
-    /// Note: descendant nodes may include some tokens that are not part of the body.
+    /// Note: descendant nodes may include some tokens that are not part of the body and exclude some tokens that are.
     /// </summary>
     public abstract OneOrMany<SyntaxNode> RootNodes { get; }
 
+    /// <summary>
+    /// <see cref="SyntaxNode"/> that includes all active tokens (<see cref="MemberBody.GetActiveTokens"/>)
+    /// and its span covers the entire <see cref="MemberBody.Envelope"/>.
+    /// May include descendant nodes or tokens that do not belong to the body.
+    /// </summary>
+    public abstract SyntaxNode EncompassingAncestor { get; }
+
     public abstract StateMachineInfo GetStateMachineInfo();
 
     /// <summary>
     /// Computes a statement-level syntax tree match of this body with <paramref name="newBody"/>.
-    /// TODO: Consider returning <see cref="BidirectionalMap{T}"/> that includes matches of all roots of the body.
-    /// <see cref="TryMatchActiveStatement"/> might not be needed if we do so.
     /// </summary>
-    public abstract Match<SyntaxNode> ComputeMatch(DeclarationBody newBody, IEnumerable<KeyValuePair<SyntaxNode, SyntaxNode>>? knownMatches);
+    public virtual BidirectionalMap<SyntaxNode> ComputeMatch(DeclarationBody newBody, IEnumerable<KeyValuePair<SyntaxNode, SyntaxNode>>? knownMatches)
```

Comment:
>ComputeMatch [](http://example.com/codeflow?start=48&length=12)

Allows us to combine matches of multiple body parts into a single bi-di map. Also allows the map to be empty.